### ISSUE TITLE
Support for UUID primary keys

### ISSUE THREAD

**Issue by dhendry** [M1ha-Shvn/django-pg-bulk-update#84]
We have been moving towards using UUID primary keys for all of our django models as a to avoid conflicts in our multi database system. Unfortunately this library (which is great by the way, thank you), does not appear to support UUID primary keys.

Consider this model:

```
class ModelWithUuidPk(models.Model):
    id = models.UUIDField(primary_key=True, default=uuid.uuid4)

    key_text = models.TextField(unique=True)
    value_text = models.TextField()
```

And then following:

```
import django_pg_bulk_update
from django.db import connection
from django.test.utils import CaptureQueriesContext

with CaptureQueriesContext(connection) as queries:
    django_pg_bulk_update.bulk_update_or_create(
        ModelWithUuidPk,
        [
            {"key_text": "foo", "value_text": "v1"},
            {"key_text": "bar", "value_text": "v2"},
        ],
        key_fields=["key_text"],
    )
```

This errors out with:

```
duplicate key value violates unique constraint "ad_engine_modelwithuuidpk_pkey"
DETAIL:  Key (id)=(b54d21cc-fd99-46e7-8293-b2db04593036) already exists.
```

Inspecting the underlying query shows:

```
WITH vals("key_0__key_text", "upd__value_text") AS (VALUES (CAST('foo' AS text), CAST('asdf' AS text)), ('bar', 'fdsa')),
 default_vals("def__id") AS (VALUES (CAST('b54d21cc-fd99-46e7-8293-b2db04593036'::uuid AS uuid))) 
        INSERT INTO "ad_engine_modelwithuuidpk" ("key_text", "value_text", "id")
        SELECT "vals"."key_0__key_text", "vals"."upd__value_text", "default_vals"."def__id" FROM "vals" CROSS JOIN "default_vals"
     ON CONFLICT ("key_text") DO UPDATE SET ("value_text") = (SELECT "vals"."upd__value_text" FROM "vals" WHERE ("key_0__key_text") = (EXCLUDED."key_text"))
```

Obviously the problem here is the use of the same default value for all rows.

This is unfortunately a blocking issue for us and there does not appear to be any easy workaround.

**Comment by M1ha-Shvn** [comment]
Hi Dan, thanks for report.

I think, that this bug is not about `UUIDField`. It is about using functions as default value generator. It works well for `AutoField` and `BigAutoField` as far as defaults are generated on PostgreSQL side by sequence, not by django.

Looks like this bug has been introduced in [v3.3.0](https://github.com/M1ha-Shvn/django-pg-bulk-update/releases/tag/v3.3.0) when I decided to optimize passing default values. Nowadays library gets them once and passes in `default_vals` query section. I did this, as query grew very large in wide models passing same default values again and again in each value tuple (firstly, I just added these values to each `vals` tuple).

I'll think what can be done here in order to fix a bug and don't send to many data at the same time. But it will require some time, as I need to dig inside django sources here.

As fast solution, you can split this query into 2, generating UUID manually:
```python
from collections import ChainMap

from django.db import connection
from django.test.utils import CaptureQueriesContext

with CaptureQueriesContext(connection) as queries:
    data = [
        {"key_text": "foo", "value_text": "v1"},
        {"key_text": "bar", "value_text": "v2"},
    ]
    
    # Insert only values that does not exist
    inserted_qs = ModelWithUuidPk.objects.pg_bulk_update_or_create(
        [ChainMap(item, {'id': uuid.uuid4()}) for item in data],
        key_fields=["key_text"], update=False, returning='id'
    )
    
    # Get inserted ids
    inserted_ids = inserted_qs.values_list('id', flat=True)
    
    # Updates fields that already existed
    ModelWithUuidPk.objects.exclude(id__in=inserted_ids).pg_bulk_update(data, key_fields=["key_text"])
```

**Comment by M1ha-Shvn** [comment]
Hi again.
I've found more elegant hotfix solution. You can define your own set function which will generate default value for insert statement and act as `EqualSetFunction` otherwise. Something like:
```python
from django_pg_bulk_update.set_functions import EqualSetFunction

class UUIDDefaultSetFunction(EqualSetFunction):
    names = {'uuid_default_function'}
    supported_field_classes = {'UUIDField'}
    needs_value = False

    def get_sql_value(self, field, val, connection, val_as_param=True, with_table=False, for_update=True,
                      **kwargs):
        if not val_as_param:
            # If field name is passed in val parameter, for instance, "vals"."id"
            return super().get_sql_value(field, val, connection, val_as_param=val_as_param,
                                         with_table=with_table, **kwargs)
                                         
        elif for_update:
            # Updates leaves field unchanged
            return self._get_field_column(field, with_table=with_table), []
            
        else:
            # Generate new UUID value here

            # If you have uuid-ossp module installed in Postgres
            return 'uuid_generate_v4()', []

            # If no module is installed
            return "%s", [field.get_default()]
```
P. s. this function ignores all values passed in `values` parameter of `bulk_update_or_create` function. It won't replace UUID if you need to update it.

**Comment by dhendry** [comment]
@M1ha-Shvn awesome, thanks a lot for both the suggestions for a short term fix and the changes it looks like you are putting together. 

I really appreciate your work on this library. Thank you!

**Comment by M1ha-Shvn** [comment]
Released [v3.5.1](https://github.com/M1ha-Shvn/django-pg-bulk-update/releases/tag/v3.5.1)
